File: 0239_Sliding Window Maximum/SlidingWindowMaximum.py
from collections import deque
class Solution:
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        if not nums:
            return []
        if k == 0:
            return nums

        indices = deque()
        ans = [0] * (len(nums)-k+1)

        for i in range(len(nums)):
            while len(indices) != 0 and nums[i] > nums[indices[-1]]:
                    indices.pop()
                
            indices.append(i)
            if (i - k + 1 >= 0):ans[i - k + 1] = nums[indices[0]]       
            if (i - k + 1 == indices[0]):indices.popleft()
        
        return ans

File: 0239_Sliding Window Maximum/SlidingWindowMaximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        if(n < k) return new int[0];
        ArrayDeque<Integer> queue = new ArrayDeque<>();
        int[] ans = new int[n - k + 1];
        int index = 0;
        
        for(int i=0;i<n;i++) {
            while(!queue.isEmpty() && i - queue.peekFirst() >= k) queue.pollFirst();
            while(!queue.isEmpty() && nums[queue.peekLast()] <= nums[i]) queue.pollLast();
            queue.addLast(i);
            if(i+1 >= k) ans[index++] = nums[queue.peekFirst()];
        }
        
        return ans;
    }
}